### PR TITLE
Rework foodstuffs sorting in the Consume Item window, accounting for sealed items

### PR DESCRIPTION
**Pull request by aerrant**
#### Summary
SUMMARY: Interface "Rework foodstuffs sorting in the Consume Item window, accounting for sealed items"

#### Purpose of change
Foodstuffs are presently sorted roughly by their perishability, but sealed foods are mixed with the rest, which is misleading. With this PR they will be sorted as follows:
1. Edible, perishable items
2. Edible, sealed items
3. Edible, non-perishable items, both sealed and not
4. Edible, rotten items
5. Inedible items (frozen, not directly edible)
With saprophage/saprovore trait the rotten food is moved to the top - everything else is the same. No other traits are handled: allergies (e.g. plant beverages for carnivores) are ignored, completely inedible (e.g. frozen solid) items are already handled elsewhere.

The original code contains a check for fresh items with 0 turns of remaining time. Don't know what it's for, didn't implement it.

This is my first PR, please be extra critical of its quality. I'm presently working on my own annoyances and there are three or more small PRs on the way.

#### Describe the solution
The comestible_inventory_preset::sort_compare() predicate in game_inventory.cpp was rewritten entirely, some helper functions were added, some removed.

#### Testing
1. Start with a generic character w/o any special skills or traits
1. Put all kinds of foodstuffs in your inventory: fresh, rotten, fresh sealed, rotten sealed, inedible.
2. Check that the sorting order is as described
3. Debug yourself a saprophage/saprovore trait
4. Check that the sorting order is the same, but now rotten items are moved on top

### CODE REVIEW

File: src/game_inventory.cpp [hunk 1]
```diff
@@ -688,30 +688,50 @@ class comestible_inventory_preset : public inventory_selector_preset
         }
 
         bool sort_compare( const inventory_entry &lhs, const inventory_entry &rhs ) const override {
-            time_duration time_a = get_time_left( lhs.any_item() );
-            time_duration time_b = get_time_left( rhs.any_item() );
-            int order_a = get_order( lhs.any_item(), time_a );
-            int order_b = get_order( rhs.any_item(), time_b );
+            // Sort by comestibility first
+            const auto lhs_comestible = lhs.any_item()->is_comestible();
+            const auto rhs_comestible = rhs.any_item()->is_comestible();
```
Comment (BrettDong) [inline src/game_inventory.cpp:693]:
```suggestion
            const bool lhs_comestible = lhs.any_item()->is_comestible();
            const bool rhs_comestible = rhs.any_item()->is_comestible();
```

File: src/game_inventory.cpp [hunk 2]
```diff
@@ -688,30 +688,50 @@ class comestible_inventory_preset : public inventory_selector_preset
...
+            if( lhs_comestible != rhs_comestible ) {
+                return !lhs_comestible && rhs_comestible;
+            }
+
+            // Then by rotten/fresh status depending on player's traits
+            const auto lhs_rotten = lhs.any_item()->rotten();
+            const auto rhs_rotten = rhs.any_item()->rotten();
```
Comment (BrettDong) [inline src/game_inventory.cpp:700]:
```suggestion
            const bool lhs_rotten = lhs.any_item()->rotten();
            const bool rhs_rotten = rhs.any_item()->rotten();
```

File: src/game_inventory.cpp [hunk 3]
```diff
@@ -688,30 +688,50 @@ class comestible_inventory_preset : public inventory_selector_preset
...
+            if( lhs_rotten != rhs_rotten ) {
+                return prefer_rotten() ^ ( !lhs_rotten && rhs_rotten );
+            }
+
+            // Then by perishability
+            const auto lhs_imperishable = !lhs.any_item()->goes_bad();
+            const auto rhs_imperishable = !rhs.any_item()->goes_bad();
```
Comment (BrettDong) [inline src/game_inventory.cpp:707]:
```suggestion
            const bool lhs_imperishable = !lhs.any_item()->goes_bad();
            const bool rhs_imperishable = !rhs.any_item()->goes_bad();
```

File: src/game_inventory.cpp [hunk 4]
```diff
@@ -688,30 +688,50 @@ class comestible_inventory_preset : public inventory_selector_preset
...
+                return !lhs_imperishable && rhs_imperishable;
+            }
 
-            return order_a < order_b
-                   || ( order_a == order_b && time_a < time_b )
-                   || ( order_a == order_b && time_a == time_b &&
-                        inventory_selector_preset::sort_compare( lhs, rhs ) );
+            // Then by sealed status, skipping imperishable items
+            const auto lhs_sealed = is_sealed( lhs.any_item() );
+            const auto rhs_sealed = is_sealed( rhs.any_item() );
```
Comment (BrettDong) [inline src/game_inventory.cpp:714]:
```suggestion
            const bool lhs_sealed = is_sealed( lhs.any_item() );
            const bool rhs_sealed = is_sealed( rhs.any_item() );
```

File: src/game_inventory.cpp [hunk 5]
```diff
@@ -688,30 +688,50 @@ class comestible_inventory_preset : public inventory_selector_preset
...
+            if( lhs_sealed != rhs_sealed && lhs.any_item()->goes_bad() ) {
+                return !lhs_sealed && rhs_sealed;
+            }
+
+            // Then by remaining freshness time, ascending
+            const auto lhs_time_left = get_time_left( lhs.any_item() );
+            const auto rhs_time_left = get_time_left( rhs.any_item() );
```
Comment (BrettDong) [inline src/game_inventory.cpp:721]:
```suggestion
            const time_duration lhs_time_left = get_time_left( lhs.any_item() );
            const time_duration rhs_time_left = get_time_left( rhs.any_item() );
```

File: src/game_inventory.cpp [hunk 6]
```diff
@@ -688,30 +688,50 @@ class comestible_inventory_preset : public inventory_selector_preset
...
+            if( lhs_time_left != rhs_time_left ) {
+                return lhs_time_left < rhs_time_left;
+            }
+
+            // Finally sort by item's generic name
+            const auto lhs_name = lhs.any_item()->type_name();
+            const auto rhs_name = rhs.any_item()->type_name();
+            return lhs_name < rhs_name;
```
Comment (BrettDong) [inline src/game_inventory.cpp:729]:
```suggestion
            return localized_compare( lhs_name, rhs_name );
```

**Comment by actual-nh** [comment]
Clang-tidy has problems with your addition of `auto` variables, and with your (fallback?) comparison on names needing to use a localized version.

**Review by BrettDong — COMMENTED** [review]
Some `clang-tidy` warnings:

File: src/game_inventory.cpp (hunk 6 above)
Comment (actual-nh) [reply src/game_inventory.cpp:729]: Question. I am trying to learn more C++. Why is `auto` OK for `lhs_name` and `rhs_name`, but problematic (or at least potentially so) for the rest? Thanks!

Comment (BrettDong) [reply src/game_inventory.cpp:729]: It's just because how rules were defined in #43370.

**Comment by aerrant** [comment]
Oh, that's embarassing. Thank you for review, actual-nh, BrettDong!

Now that I think of it, I'll add more sorting clauses, for frozen items and (maybe) for items made inedible by mutations.

**Comment by actual-nh** [comment]
> Now that I think of it, I'll add more sorting clauses, for frozen items and (maybe) for items made inedible by mutations.

Health effect?

File: src/game_inventory.cpp [hunk 7]
```diff
@@ -688,30 +688,50 @@ class comestible_inventory_preset : public inventory_selector_preset
...
+                return !lhs_imperishable && rhs_imperishable;
+            }
 
-            return order_a < order_b
-                   || ( order_a == order_b && time_a < time_b )
-                   || ( order_a == order_b && time_a == time_b &&
-                        inventory_selector_preset::sort_compare( lhs, rhs ) );
+            // Then by sealed status, skipping imperishable items
+            const bool lhs_sealed = is_sealed( lhs.any_item() );
+            const bool rhs_sealed = is_sealed( rhs.any_item() );
```
Comment (Qrox) [inline src/game_inventory.cpp:714]: Items can still rot in a sealed pocket if the pocket has a non-zero spoil multiplier (for example plastic bottles have a sealed spoil multiplier of 0.5). Addtionally unsealed pockets can also have a spoil multiplier less than 1 (for example condiment bottles have a unsealed spoil multiplier of 0.5).

I suggest to just compare the time left divided by the spoil multiplier (and use multiplication for numerical stability, i.e. `lhs_time_left * rhs_spoil_multiplier < rhs_time_left * lhs_spoil_multiplier`).

**Comment by actual-nh** [comment]
You can add "Fixes #46866", BTW.

**Comment by Night-Pryanik** [comment]
@aerrant is this being worked on?

**Comment by I-am-Erk** [comment]
Closing as abandoned, but I think this is something we want so I bookmarked it as such.

**Pull request closed without merging**